### ISSUE TITLE
ingress-controller out of sync/lost config after ingressClass object replacement

### ISSUE THREAD

**Issue by kirrmann** [haproxytech/kubernetes-ingress#412]
As I prepared the update from helm chart 1.16.2 to 1.17.11 I noticed that both charts setup ingressClass differently. 1.16.2 setting it up kind of wrong, which is addressed by 1.17.11 which allows setting the controller name explicitly. This results is a breaking change because the controller field being immutable. So I tried to replace the ingressClass before doing a helm upgrade, but randomly hitting 404 for some pods after doing so.

I can trigger the issue by:
1. helm chart 1.16.2 has ingress-controller version 1.6.5 with 3 pods deployed
2. debug logs enabled, ingress.class is set, sync-period is set to default and cache-resync-period to 10s
2. I'm running siege in another console to check for http response codes
3. I'm replacing the ingressClass with `kubectl get ingressClass haproxy-a -o json | jq '.spec.controller = "haproxy.org/ingress-controller/haproxy-a"' | kubectl replace --force -f -` original it's `.spec.controller = "haproxy.org/ingress-controller`
4. debug logs (`| grep -i debug`)
```
haproxy-a-ingress-kubernetes-ingress-56d774d5d5-ssvvs kubernetes-ingress-controller 2022/01/13 14:08:51 DEBUG   controller.go:156 ingress 'echoserver/echoserver' ignored: no matching IngressClass
haproxy-a-ingress-kubernetes-ingress-56d774d5d5-4lgzc kubernetes-ingress-controller 2022/01/13 14:08:52 DEBUG   controller.go:156 ingress 'echoserver/echoserver' ignored: no matching IngressClass
haproxy-a-ingress-kubernetes-ingress-56d774d5d5-4lgzc kubernetes-ingress-controller 2022/01/13 14:08:52 DEBUG   haproxy/maps.go:116 Map file 'host' updated, reload required
haproxy-a-ingress-kubernetes-ingress-56d774d5d5-4lgzc kubernetes-ingress-controller 2022/01/13 14:08:52 DEBUG   haproxy/maps.go:116 Map file 'path-prefix' updated, reload required
haproxy-a-ingress-kubernetes-ingress-56d774d5d5-4lgzc kubernetes-ingress-controller 2022/01/13 14:08:52 DEBUG   handler/refresh.go:56 Deleting backend 'echoserver-echoserver-80'
```

-> pod 4lgzc deletes the backend
-> pod ssvvs only outputs `no matching IngressClass`
-> pod 59jjq not outputting any debug log

-> pod 4lgzc immediately starts returning 404s
-> pod ssvvs returns 200s but after a while starts returning 404s
-> pod 59jjq keeps returning 200s

debug logs don't show anything happening after this and the issue doesn't resolve on its own. I continue replacing the ingressClass multiple times I could get lucky and all pods are returning 200s again.

Almost the same behaviour I can see in ingress-controller 1.7.3, which is setup in the same way, but with this version the issue resolves itself seeing following logs

```
haproxy-a-ingress-kubernetes-ingress-66b655f486-2pt85 kubernetes-ingress-controller 2022/01/13 14:59:56 DEBUG   controller.go:168 ingress 'echoserver/echoserver' ignored: no matching IngressClass
haproxy-a-ingress-kubernetes-ingress-66b655f486-2pt85 kubernetes-ingress-controller 2022/01/13 14:59:56 DEBUG   maps/main.go:124 Map file 'host' updated, reload required
haproxy-a-ingress-kubernetes-ingress-66b655f486-2pt85 kubernetes-ingress-controller 2022/01/13 14:59:56 DEBUG   maps/main.go:124 Map file 'path-prefix' updated, reload required
haproxy-a-ingress-kubernetes-ingress-66b655f486-2pt85 kubernetes-ingress-controller 2022/01/13 14:59:56 DEBUG   handler/refresh.go:57 Deleting backend 'echoserver-echoserver-80'
haproxy-a-ingress-kubernetes-ingress-66b655f486-2pt85 kubernetes-ingress-controller 2022/01/13 15:00:01 DEBUG   service/service.go:130 Service 'echoserver/echoserver': new backend 'echoserver-echoserver-80', reload required
haproxy-a-ingress-kubernetes-ingress-66b655f486-2pt85 kubernetes-ingress-controller 2022/01/13 15:00:01 DEBUG   maps/main.go:124 Map file 'host' updated, reload required
haproxy-a-ingress-kubernetes-ingress-66b655f486-2pt85 kubernetes-ingress-controller 2022/01/13 15:00:01 DEBUG   maps/main.go:124 Map file 'path-prefix' updated, reload required
```

Maybe worth mentioning it could be that during my siege not all pods were hit by the same amount of requests, but none of them exhausted their resources.

**Comment by ivanmatmati** [comment]
Hi, controller v1.6.5 does not support the extended mechanism in ingressclass you can use in v1.7.x. Yet, we have found a bug  that could explain the delay in updating you saw with version 1.7.3. Please use the latest commit (7c55137c31c019f309c15c731a7c1009cd272a6f) to test the fix. For v1.6.5, we might backport the extended mechanism in a future release of v1.6.

**Comment by kirrmann** [comment]
Hi, sorry it took me a while to test this

I deployed the commit in our cluster and after the pods were healthy I replaced the ingressClass object. But as soon as I did this I was hitting 404s, which then resolve itself after 10 seconds or so

Log output:
```
kubernetes-ingress-controller 10.129.5.12:61928 [15/Feb/2022:11:29:51.833] http echoserver_echoserver_80/SRV_1 0/0/0/1/1 200 498 - - ---- 4/1/0/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller 10.129.5.12:61929 [15/Feb/2022:11:29:51.841] http echoserver_echoserver_80/SRV_1 0/0/1/1/2 200 498 - - ---- 5/2/0/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller 2022/02/15 11:29:51 ERROR   global.go:171 default certificate: annotation 'ssl-certificate': secret 'haproxy-nlb-ip-ingress-test/haproxy-nlb-ip-ingress-test-kubernetes-ingress-default-cert' does not exist
kubernetes-ingress-controller 2022/02/15 11:29:51 DEBUG   controller.go:168 ingress 'echoserver/echoserver' ignored: no matching IngressClass
kubernetes-ingress-controller 2022/02/15 11:29:51 DEBUG   maps/main.go:124 Map file 'path-prefix' updated, reload required
kubernetes-ingress-controller 10.129.5.12:61926 [15/Feb/2022:11:29:51.844] http echoserver_echoserver_80/SRV_1 0/0/0/2/2 200 498 - - ---- 4/1/0/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller 2022/02/15 11:29:51 DEBUG   maps/main.go:124 Map file 'host' updated, reload required
kubernetes-ingress-controller 2022/02/15 11:29:51 DEBUG   handler/refresh.go:57 Deleting backend 'echoserver_echoserver_80'
kubernetes-ingress-controller 10.129.5.12:61931 [15/Feb/2022:11:29:51.852] http echoserver_echoserver_80/SRV_1 0/0/1/1/2 200 498 - - ---- 5/2/1/1/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller 10.129.5.12:61930 [15/Feb/2022:11:29:51.853] http echoserver_echoserver_80/SRV_1 0/0/0/1/1 200 498 - - ---- 4/1/0/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller 10.129.5.12:61932 [15/Feb/2022:11:29:51.857] http echoserver_echoserver_80/SRV_1 0/0/0/1/1 200 498 - - ---- 5/2/1/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller 10.129.5.12:61933 [15/Feb/2022:11:29:51.858] http echoserver_echoserver_80/SRV_1 0/0/1/1/2 200 498 - - ---- 4/1/0/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller 10.129.5.12:61937 [15/Feb/2022:11:29:51.868] http echoserver_echoserver_80/SRV_1 0/0/1/1/2 200 498 - - ---- 5/2/1/1/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller 2022/02/15 11:29:51 INFO    controller.go:211 HAProxy reloaded
kubernetes-ingress-controller 10.129.5.12:61935 [15/Feb/2022:11:29:51.870] http echoserver_echoserver_80/SRV_1 0/0/0/2/2 200 498 - - ---- 4/1/0/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller [WARNING]  (257) : Reexecuting Master process
kubernetes-ingress-controller 10.129.5.12:61938 [15/Feb/2022:11:29:51.874] http echoserver_echoserver_80/SRV_1 0/0/2/1/3 200 498 - - ---- 5/2/0/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller 10.129.5.12:61939 [15/Feb/2022:11:29:51.878] http echoserver_echoserver_80/SRV_1 0/0/0/1/1 200 498 - - ---- 4/1/0/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller [WARNING]  (257) : config: Can't get version of the global server state file '/var/state/haproxy/global'.
kubernetes-ingress-controller 10.129.5.12:61934 [15/Feb/2022:11:29:51.885] http echoserver_echoserver_80/SRV_1 0/0/1/1/2 200 498 - - ---- 5/2/1/1/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller 10.129.5.12:61936 [15/Feb/2022:11:29:51.885] http echoserver_echoserver_80/SRV_1 0/0/1/1/2 200 498 - - ---- 4/1/0/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller Proxy healthz stopped (cumulated conns: FE: 20, BE: 0).
kubernetes-ingress-controller Proxy http stopped (cumulated conns: FE: 1832, BE: 0).
kubernetes-ingress-controller Proxy https stopped (cumulated conns: FE: 0, BE: 0).
kubernetes-ingress-controller Proxy stats stopped (cumulated conns: FE: 6, BE: 0).
kubernetes-ingress-controller [WARNING]  (270) : Proxy healthz stopped (cumulated conns: FE: 20, BE: 0).
kubernetes-ingress-controller [WARNING]  (270) : Proxy http stopped (cumulated conns: FE: 1832, BE: 0).
kubernetes-ingress-controller [WARNING]  (270) : Proxy https stopped (cumulated conns: FE: 0, BE: 0).
kubernetes-ingress-controller [WARNING]  (270) : Proxy stats stopped (cumulated conns: FE: 6, BE: 0).
kubernetes-ingress-controller [WARNING]  (270) : Stopping frontend GLOBAL in 0 ms.
kubernetes-ingress-controller [WARNING]  (270) : Stopping backend echoserver_echoserver_80 in 0 ms.
kubernetes-ingress-controller [WARNING]  (270) : Stopping backend haproxy-nlb-ip-ingress-test_haproxy-nlb-ip-ingress-test-kubernetes-ingress-default-backend_http in 0 ms.
kubernetes-ingress-controller Stopping backend echoserver_echoserver_80 in 0 ms.
kubernetes-ingress-controller Stopping backend haproxy-nlb-ip-ingress-test_haproxy-nlb-ip-ingress-test-kubernetes-ingress-default-backend_http in 0 ms.
kubernetes-ingress-controller [NOTICE]   (257) : New worker #1 (280) forked
kubernetes-ingress-controller 10.129.5.12:61943 [15/Feb/2022:11:29:51.911] http haproxy-nlb-ip-ingress-test_haproxy-nlb-ip-ingress-test-kubernetes-ingress-default-backend_http/SRV_2 0/0/1/0/1 404 148 - - ---- 5/5/4/1/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller 10.129.5.12:61944 [15/Feb/2022:11:29:51.911] http haproxy-nlb-ip-ingress-test_haproxy-nlb-ip-ingress-test-kubernetes-ingress-default-backend_http/SRV_2 0/0/0/2/2 404 148 - - ---- 4/4/3/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
...
...
kubernetes-ingress-controller 10.129.5.12:63050 [15/Feb/2022:11:29:56.845] http haproxy-nlb-ip-ingress-test_haproxy-nlb-ip-ingress-test-kubernetes-ingress-default-backend_http/SRV_2 0/0/1/0/1 404 148 - - ---- 2/2/1/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller 2022/02/15 11:29:56 ERROR   global.go:171 default certificate: annotation 'ssl-certificate': secret 'haproxy-nlb-ip-ingress-test/haproxy-nlb-ip-ingress-test-kubernetes-ingress-default-cert' does not exist
kubernetes-ingress-controller 2022/02/15 11:29:56 DEBUG   service/service.go:130 Service 'echoserver/echoserver': new backend 'echoserver_echoserver_80', reload required
kubernetes-ingress-controller 10.129.5.12:63054 [15/Feb/2022:11:29:56.845] http haproxy-nlb-ip-ingress-test_haproxy-nlb-ip-ingress-test-kubernetes-ingress-default-backend_http/SRV_1 0/0/1/1/2 404 148 - - ---- 1/1/0/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller 2022/02/15 11:29:56 DEBUG   maps/main.go:124 Map file 'host' updated, reload required
kubernetes-ingress-controller 10.129.5.12:63060 [15/Feb/2022:11:29:56.849] http haproxy-nlb-ip-ingress-test_haproxy-nlb-ip-ingress-test-kubernetes-ingress-default-backend_http/SRV_2 0/0/1/0/1 404 148 - - ---- 1/1/0/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller 2022/02/15 11:29:56 DEBUG   maps/main.go:124 Map file 'path-prefix' updated, reload required
kubernetes-ingress-controller 2022/02/15 11:29:56 INFO    controller.go:211 HAProxy reloaded
kubernetes-ingress-controller [WARNING]  (257) : Reexecuting Master process
kubernetes-ingress-controller 10.129.5.12:63061 [15/Feb/2022:11:29:56.877] http haproxy-nlb-ip-ingress-test_haproxy-nlb-ip-ingress-test-kubernetes-ingress-default-backend_http/SRV_1 0/0/0/1/1 404 148 - - ---- 1/1/0/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller [WARNING]  (257) : config: Can't get version of the global server state file '/var/state/haproxy/global'.
kubernetes-ingress-controller 10.129.5.12:63064 [15/Feb/2022:11:29:56.888] http haproxy-nlb-ip-ingress-test_haproxy-nlb-ip-ingress-test-kubernetes-ingress-default-backend_http/SRV_2 0/0/0/1/1 404 148 - - ---- 2/2/1/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller 10.129.5.12:63062 [15/Feb/2022:11:29:56.888] http haproxy-nlb-ip-ingress-test_haproxy-nlb-ip-ingress-test-kubernetes-ingress-default-backend_http/SRV_1 0/0/1/1/2 404 148 - - ---- 1/1/0/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller 10.129.5.12:63065 [15/Feb/2022:11:29:56.895] http haproxy-nlb-ip-ingress-test_haproxy-nlb-ip-ingress-test-kubernetes-ingress-default-backend_http/SRV_2 0/0/1/0/1 404 148 - - ---- 1/1/0/0/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller Proxy healthz stopped (cumulated conns: FE: 0, BE: 0).
kubernetes-ingress-controller Proxy http stopped (cumulated conns: FE: 1104, BE: 0).
kubernetes-ingress-controller Proxy https stopped (cumulated conns: FE: 0, BE: 0).
kubernetes-ingress-controller Proxy stats stopped (cumulated conns: FE: 0, BE: 0).
kubernetes-ingress-controller Stopping backend haproxy-nlb-ip-ingress-test_haproxy-nlb-ip-ingress-test-kubernetes-ingress-default-backend_http in 0 ms.
kubernetes-ingress-controller [WARNING]  (280) : Proxy healthz stopped (cumulated conns: FE: 0, BE: 0).
kubernetes-ingress-controller [WARNING]  (280) : Proxy http stopped (cumulated conns: FE: 1104, BE: 0).
kubernetes-ingress-controller [WARNING]  (280) : Proxy https stopped (cumulated conns: FE: 0, BE: 0).
kubernetes-ingress-controller [WARNING]  (280) : Proxy stats stopped (cumulated conns: FE: 0, BE: 0).
kubernetes-ingress-controller [NOTICE]   (257) : New worker #1 (288) forked
kubernetes-ingress-controller [WARNING]  (280) : Stopping frontend GLOBAL in 0 ms.
kubernetes-ingress-controller [WARNING]  (280) : Stopping backend haproxy-nlb-ip-ingress-test_haproxy-nlb-ip-ingress-test-kubernetes-ingress-default-backend_http in 0 ms.
kubernetes-ingress-controller 10.129.5.12:63068 [15/Feb/2022:11:29:56.903] http echoserver_echoserver_80/SRV_1 0/0/1/1/2 200 498 - - ---- 8/8/7/7/0 0/0 "GET echo.example.com/ HTTP/1.1"
kubernetes-ingress-controller 10.129.5.12:63073 [15/Feb/2022:11:29:56.903] http echoserver_echoserver_80/SRV_1 0/0/1/2/3 200 498 - - ---- 7/7/6/6/0 0/0 "GET echo.example.com/ HTTP/1.1"

```

**Comment by kirrmann** [comment]
Yes that's true, sync period was set to 5s (default)

**Comment by kirrmann** [comment]
Seems this is covered by https://github.com/haproxytech/kubernetes-ingress/commit/07e10cdfa63e3cdf179a9b8eaaadac966613ff04 which is released in v1.7.7

**Comment by Mo3m3n** [comment]
Thanks for your feedback @kirrmann 
Feel free to re-open this issue if the the problem reappears.
